docs/CODE_GUIDELINES.md

```diff
@@ -147,6 +151,16 @@ else
   return;
 ```
 
+❌ Bad:
+```cpp
+if (true)
+  return; // No new line after if statement
```

Comment:
The rule doesn't refer to this. It also seems to be pedantic. I agree whitespace is good but I don't think we need a hard rule here.

Reply:
Indeed, this covers the topic of vertical whitespace, and that's more of a readability issue than a style issue. Maybe mention that vertical whitespace can create "stanzas" of code for the reader to digest.

Reply:
i moved that part to `3.4. Whitespace` in a different formula

this reminded me of something else
we allow the use of control statements without brackets?
e.g.

```
[..........other code....]
while (....)
  x+=1;
[..........other code....]
```

vs:

```
[..........other code....]
while (....)
{
  x+=1;
}
[..........other code....]
```

I don't seem to have found any mention of it
I personally do not like the use of them without brackets, opinions?

Reply:
for me this is fine:

```
[..........other code....]
while (....)
  x+=1;
                              <--- blank line
[..........other code....]
```

Reply:
i think i have wrote a bad example, my fear is that we allow something this:

```c++
    [..........other code....]
    for (std::vector<AdaptationSet*>::const_iterator ba(current_period_->adaptationSets_.begin()),
         ea(current_period_->adaptationSets_.end());
         ba != ea; ++ba)
      for (std::vector<Representation*>::iterator br((*ba)->representations_.begin()),
           er((*ba)->representations_.end());
           br != er; ++br)
        if ((*br)->flags_ & Representation::ENABLED)
          refresh_list.push_back(std::make_tuple(*ba, *br));
    [..........other code....]
```

on InputStreamAdaptive addon i cleaned up a lot of code like this, which made the code so difficult to read especially when the block of code was long

on this case the "for" condition is very long, and become splitted on more lines, plus nested control statements without brackets make all this hard to read

Reply:
sorry it's untested but arent they perfect candidates for range based for ? or some other other reason that doesnt allow
```
 [..........other code....]
    for (const auto& ba : current_period_->adaptationSets_())
      for (const auto& br : ba->representations_())
        if (br->flags_ & Representation::ENABLED)
          refresh_list.push_back(std::make_tuple(ba, br));
[..........other code....]
```

Reply:
was only and example you can use any type of content in the conditions
that become splitted on more lines when you clang-format
i find it quite a mess

Reply:
so my last concern is the use/abuse of `auto` expression
its clear that auto can be used everywhere without problem because the type is deduced by value by compiler
but i think this becomes a bit risky and difficult to read at least when there are more variables

in some circustances like this example:
https://github.com/xbmc/xbmc/pull/22705/files#r1096691684
we declare a list of constants variables
the types not being explicit set a little oversight can declare a wrong type
and when you have to investigate reading all `auto` it is not so immediate to understand each type

we can think to add in doc some kind of limitation for this use case, maybe for fundamental types only?

Reply:
One note here: this code is automatically generated by Visual Studio 2022 "hints" when suggest change #defines to `constexpr` maybe it's not such a bad thing.

Usually auto is used on very long types or complex or having multiple sub-levels in a namespace.
Also seems can be used when type is easily deduced from name of variable (e.g. type is a class).

```c++
auto hdmiInfo = HdmiDisplayInformation::GetForCurrentView();
```

Reply:
> suggest change #defines to constexpr maybe it's not such a bad thing.

i think its not in docs because new code should not introduce new #defines

> Usually auto is used on very long types or complex or having multiple sub-levels in a namespace.

yes for this use case i agree